src/decide.ts

```diff
@@ -102,6 +104,48 @@ export function compareFileTreesIgnoringNames(
 	);
 }
 
+export function compareFileTreesPartialIgnoringNames(
+	candidate: Metafile,
+	searchee: Searchee
+): boolean {
+	const { fuzzySizeThreshold } = getRuntimeConfig();
+	let matchedSizes = 0;
+	for (const candidateFile of candidate.files) {
+		const searcheeHasFileSize = searchee.files.some(
+			(searcheeFile) => searcheeFile.length === candidateFile.length
+		);
+		if (searcheeHasFileSize) {
+			matchedSizes += candidateFile.length;
+		}
+	}
+	return matchedSizes / candidate.length >= 1 - fuzzySizeThreshold;
+}
+
+export function compareFileTreesPartial(
+	candidate: Metafile,
+	searchee: Searchee
+): boolean {
+	if (candidate.length <= candidate.pieceLength) {
+		return false;
+	}
```

Comment:
I still don't really understand this bit. what makes it not work when the piece length is too big?

Reply:
piece length shouldn't have any issue with the data matching unless the candidate is not a complete torrent....i also do not understand what's going on with the piece stuff.

Reply:
I've had about 10 ebooks all with 1mb piece size and torrent size less than that. They all are all stuck at 0 after rechecking. Some torrents which are between one and two pieces are above 50% but below the expected amount. Maybe all these happen to be different with the exact same size but there is some weirdness here.

In fact most partial torrents never seem to download exactly what's needed. They all download a bit more, usually one or two pieces or something in between if qbit stats are to be trusted.

The ebooks are all public so I can provide them so you guys can test.

Reply:
They sound like false positives by what you're saying. The files would otherwise be identical regardless of the piece size, and that's what counts during rechecking.

I made a post, [here](https://forum.deluge-torrent.org/viewtopic.php?p=238007&hilit=pieces#p238007) - the second sentence explains how pieces function relative to the torrent file process. Might help clarify.

I'd ask you to download both of the "matches" separately and run `sha1sum` on both files. Those 10 ebooks with 1MB pieces size that are all less than 1MB would all be 1 piece. If the `sha1sum` of that piece is different, based on rechecking to 0. Thus false-positive. Torrent checking/files use sha1. 

I'd expect the others to be much the same.

Reply:
So after discussions with @ShanaryS in discord, he showed me what was happening. This is the basic "worst-case" replicable steps.

1. Get a torrent downloaded with a size less than the piece length, quantity of files is irrelevant but 1 makes it easiest.
2. Find a torrent with that same file, and an additional (txt file, for example, i common)
3. add that second torrent (this would be the candidate) to your client.
4. because the entire piece fails in the candidate during sha1 checking, it redownloads the entire piece
5. this effectively makes cross-seeding pointless, as you downloaded the entire torrent again anyway for the missing txt file

This has less effect, but still some, on somewhat larger torrents, but becomes more and more insignificant the larger the torrent is relative to the total size of the torrent you are cross-seeding. It can also be exacerbated by "unlucky" piece distribution or cutoff points.

He has an approach he is going to use (change) that may help to address this incrementally.

Reply:
The fix I implemented is to use the ratio of missing pieces rather than missing size instead. This handles torrents with low pieces well. But it will also match less torrents with large piece sizes. For example a 320mb torrent with 16mb piece size requires a 0.05 fuzzy but it may only be missing a file that's a few bytes. Previously that could match with pretty much any non-zero fuzzy but ended up downloading the entire 16mb piece. So the new logic better estimates what needs to be downloaded.

Now there is still a chance that the file being spread across multiple pieces could cause more to be downloaded that expected but we can't know that until it's rechecked. Assuming they have skipRecheck to false the user has to be the one to start it so I think it's fine.